### PR TITLE
[Terrain] Height blending / bug fixes

### PR DESCRIPTION
**Pull request by invertednormal**
This adds optional height blending to terrain detail materials and includes a variety of bug fixes.

Height blending can be turned on in the terrain rendering component. While pom heights on the StandardPbr material can be used for height information, it's recommended to use the new TerrainDetailMaterial material type for terrain detail materials. This material type exposes additional parameters to have fine grain control over height based blending without affecting POM settings.

image

Additional options are now available in terrain components. 

The surface data resolution can now be set independently of the height query resolution:
image

The terrain renderer component has options for tweaking the render distance, fade out, and scale of the detail material:
image

Material changes:
- Lua functors can now use material properties without having a hard dependency on them. They can query for the existence of a property with `HasMaterialProperty`. This will allow lua functors to be more flexible and be used with multiple similar material types.
- Fixed an issue with importing material json snippets across gem boundaries

Bug fixes:
- Sometimes on initial load, terrain detail materials wouldn't be visible until some change was made. This should no longer happen.
- Query resolutions that didn't evenly divide the world bounds of terrain would often result in a crash, this is now supported
- Reloading the terrain material or shader would usually result in turning all the terrain metallic, this should no longer happen.

### CODE REVIEW

File: Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype
Comment (mbalfour-amzn) [inline Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype:26]: This looks a little bit sketchy, as it assumes the location of the Terrain Gem vs the Atom Gem. I won't block the PR over it since both gems are currently distributed together in the engine, but is there a better way to describe the base relative path to look this up from? like "@GEMROOT:Atom@" or something?

File: Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype
Comment (mbalfour-amzn) [inline Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype:60]: "mateirals" -> "materials"

File: Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype
Comment (mbalfour-amzn) [inline Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype:69]: Can this be split into 2 lines? It's pretty long.

File: Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype
Comment (mbalfour-amzn) [inline Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype:93]: Same here - "@GEMROOT:Atom@"?

File: Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailHelpers.azsli
Comment (mbalfour-amzn) [inline Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailHelpers.azsli:353]: This could really use some comments to help future code maintainers. In particular, an explanation of how to make a new one of these would be nice. :)

File: Gems/Terrain/Code/Source/Components/TerrainWorldRendererComponent.cpp
Comment (mbalfour-amzn) [inline Gems/Terrain/Code/Source/Components/TerrainWorldRendererComponent.cpp:47]: "materail" -> "material"

File: Gems/Terrain/Code/Source/Components/TerrainWorldRendererComponent.cpp
Comment (mbalfour-amzn) [inline Gems/Terrain/Code/Source/Components/TerrainWorldRendererComponent.cpp:53]: "materails" -> "materials"

File: Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype
Comment (invertednormal) [reply Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype:60]: material spelling strikes again! 😝

File: Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype
Comment (invertednormal) [reply Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype:69]: I don't believe that's allowed in json.

File: Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype
Comment (invertednormal) [reply Gems/Terrain/Assets/Materials/Terrain/TerrainDetailMaterial.materialtype:26]: Actually I think @santorac emailed about this to try to figure out a better solution and we may be able to do use a callback to resolve the path to just Materials/Types/etc... we can do that later though.

File: Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailHelpers.azsli
Comment (invertednormal) [reply Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailHelpers.azsli:353]: Generating an optimal sort network is hard, and optimal ones are only known for small numbers of compares. You can optimize for depth (for parallel processing) or total number of compares. It's not quite as simple as "run quicksort and record the comparisons" ... that'll get you something decent but not optimal. I found it quite interesting that you can sort 8 items with 19 compares since 8*log2(8) = 24, so the sort network is about 20% faster. See https://en.wikipedia.org/wiki/Sorting_network for an overview. I used https://demonstrations.wolfram.com/SortingNetworks/ to generate the network for 8 compares. I'll add those links in comments.

File: Gems/Terrain/Code/Source/Components/TerrainWorldRendererComponent.cpp
Comment (invertednormal) [reply Gems/Terrain/Code/Source/Components/TerrainWorldRendererComponent.cpp:53]: 😭

**Review by mbalfour-amzn — APPROVED** [review]

**Review by cgalvan — APPROVED** [review]

File: Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailHelpers.azsli
Comment (jiaweig-amzn) [inline Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailHelpers.azsli:392]: We have GetDefaultDetailSurface, might be a better fit.

File: Gems/Terrain/Assets/Shaders/Terrain/TerrainSrg.azsli
Comment (invertednormal) [reply Gems/Terrain/Assets/Shaders/Terrain/TerrainSrg.azsli:54]: A struct in hlsl is always going to be 16 byte aligned, so these aren't even technically necessary - their inclusion is to make the members match up with the c++ side to avoid any confusion.

File: Gems/Terrain/Code/Source/TerrainRenderer/TerrainFeatureProcessor.cpp
Comment (invertednormal) [reply Gems/Terrain/Code/Source/TerrainRenderer/TerrainFeatureProcessor.cpp:229]: I brought this up with @mbalfour-amzn, and the reason for the size_t is it's also common to compare these values to size_t types like AZStd::vector::size(). Any way we go about it we're going to need to do some casting. I'd also prefer this to be a uint32_t but it's ok as a size_t too.

File: Gems/Terrain/Code/Source/TerrainRenderer/TerrainFeatureProcessor.cpp
Comment (invertednormal) [reply Gems/Terrain/Code/Source/TerrainRenderer/TerrainFeatureProcessor.cpp:272]: Will fix, some tabs snuck in - new visual studio install didn't have the white space settings set up correctly.

File: Gems/Atom/RPI/Code/Source/RPI.Edit/Material/LuaMaterialFunctorSourceData.cpp
Comment (invertednormal) [reply Gems/Atom/RPI/Code/Source/RPI.Edit/Material/LuaMaterialFunctorSourceData.cpp:211]: I checked this with @santorac and he agreed with relaxing this. If you go the route of new functor every time you have a slightly different material type, the number of functors will explode - then any time anything changes that affects all the functors, you need to fix all of them.

There's no real reduction in debugging visibility. With this change you will still get errors if the functor tries to access a property that doesn't exist. This just allows the functor to intentionally be flexible about specific properties.

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp
Comment (invertednormal) [reply Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp:454]: The bug is that the path is relative from a different location in the create jobs phase vs this material building phase (source asset path vs cache path), and it was impossible to create a relative path that worked in both scenarios. This made it impossible for relative paths to work across gems.

File: Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailHelpers.azsli
Comment (invertednormal) [reply Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailHelpers.azsli:392]: This actually needs to be zeroed out. Later, specific detail materials will be add their data to the surface, so you need to start at 0 for that to work correctly.

File: Gems/Atom/RPI/Code/Source/RPI.Edit/Material/LuaMaterialFunctorSourceData.cpp
Comment (jiaweig-amzn) [reply Gems/Atom/RPI/Code/Source/RPI.Edit/Material/LuaMaterialFunctorSourceData.cpp:211]: Ok. That sounds good to me.

**Review by jiaweig-amzn — APPROVED** [review]

**Pull request merged**
